cc_growpart: Add realpath translation and adjust logging

raharper opened the pull request:
Growpart has its own internal logic for selecting which underlying
tools to use to modify the target device partition table.  This
information is not visible in cloud-init logging which makes it difficult
to understand what may have occurred when debugging.  Adding the
'--verbose' flag and logging the output will emit useful information
about which tool was used (sfdisk vs. sgdisk).

On some platforms the target device may point to a symbolic link,
e.g. /dev/disk/by-partuuid/XXXX;  These symbolic links may be
unavailable during partition changes due to udev behavior.  Use
os.path.realpath to expand the symbolic links to the real device
path (e.g. /dev/sda1); note this is already done when cloud-init
calculates the path to the partition.

ResizeGrowpart will now parse the output from growpart to report
the before and after size change (or no change at all) after running
growpart.

LP: #1834875

mwhudson commented:
> Growpart has its own internal logic for selecting which underlying
> tools to use to modify the target device partition table. This
> information is not visible in cloud-init logging which makes it difficult
> to understand what may have occurred when debugging. Adding the
> '--verbose' flag and logging the output will emit useful information
> about which tool was used (sfdisk vs. sgdisk).

This sounds useful.

> On some platforms the target device may point to a symbolic link,
> e.g. /dev/disk/by-partuuid/XXXX; These symbolic links may be
> unavailable during partition changes due to udev behavior. Use
> os.path.realpath to expand the symbolic links to the real device
> path (e.g. /dev/sda1) as this value use used in the get_size() method
> cc_growpart uses to compare before and after device values.

This sounds more or less sensible too.

> Lastly get_size() will now make use of flock (1) to take a shared
> lock against the device it will attempt to obtain the size preventing
> cc_growpart from reading the size value while udev or other processes
> are making changes.

This, however. What's the point? Taking the lock around just the read doesn't prevent a TOCTOU. Mentioning udev here also seems misleading because (a) udev doesn't modify the disk (b) udev takes a shared lock too...

raharper commented:
> > Lastly get_size() will now make use of flock (1) to take a shared
> > lock against the device it will attempt to obtain the size preventing
> > cc_growpart from reading the size value while udev or other processes
> > are making changes.
> 
> This, however. What's the point? Taking the lock around just the read doesn't prevent a TOCTOU. Mentioning udev here also seems misleading because (a) udev doesn't modify the disk (b) udev takes a shared lock too...

Udev is the wrong program here w.r.t the race.  However, I do think it is reasonable to take the shared lock as a reader of the device just as udev is doing.  That said, udev holds the lock while executing arbitrary scripts which may take considerable time; significantly more time than a read.

With a fixed growpart (or sfdisk using locking, etc) once growpart has exited, we can expect the size to be correct.  I think I'll replace get_size() with a read from sysfs instead;  that's lighter weight anyhow.

smoser commented:
This seems mostly "knee jerk" response, adding some code and extensive logging which has never been needed in 6 years.

The following statement seems to indicate to me that there will still be a bug
in growpart after the suggested fix there:

 | On some platforms the target device may point to a symbolic link,
 | e.g. /dev/disk/by-partuuid/XXXX; These symbolic links may be
 | unavailable during partition changes due to udev behavior.

After invoking 'growpart /dev/device part-number', the caller should
expect that everything is finished. The caller should not have to
wait for further events to  make sure device links are present.
Anything else is non-sense being pushed up a chain.
Growpart should do the right thing.

Secondly, I may be reading this wrong, but the added complexity of flock
is just to improve the log message.  I agree that this is useful
information.  Growpart writes (mostly) that information to stdout.
Instead of reading it again, lets parse the output of the program
that did the change:

   CHANGED: partition=1 start=2048 old: size=1021952 end=1024000 new: size=2045919,end=2047967

It sucks that this output could be better formed.  And that it is missing
sector size, but it is definitely parseable with a fairly simple regex.
I'd suggest we add sector size to growpart output, fix the lack of a ',' 
between old size/new, and accept either one here.

Third, if you're going to use flock, please just use fctn.flock which
is in the standard library.

smoser commented:
oh. and now looking at growpart it seems that somewhere between bionic (running on my system) and current cloud-utils master, we've modified the output to be consistent.
(cloud-utils commit 099bdf30b9c73c6924e632e174a5e988972301df).

so cloud-utils in bionic shows:

    CHANGED: partition=1 start=2048 old: size=1021952 end=1024000 new: size=2045919,end=2047967

cloud-utiuls in master shows:

    CHANGED: resizer=sfdisk partition=1 start=2048 old: size=1021952 end=1024000 new: size=2045919 end=2047967

raharper commented:
> This seems mostly "knee jerk" response, adding some code and extensive logging which has never been needed in 6 years.

It seems that things have changed, whether it be systemd, udev, locking, sfdisk, etc.  If we had encountered a bug in the path, we would have wished we had better logging in cc_growpart to understand what code actually ran.

> 
> The following statement seems to indicate to me that there will still be a bug
> in growpart after the suggested fix there:
> 
> | On some platforms the target device may point to a symbolic link,
> | e.g. /dev/disk/by-partuuid/XXXX; These symbolic links may be
> | unavailable during partition changes due to udev behavior.
> 
> After invoking 'growpart /dev/device part-number', the caller should
> expect that everything is finished. The caller should not have to
> wait for further events to make sure device links are present.
> Anything else is non-sense being pushed up a chain.
> Growpart should do the right thing.

Yes, I agree; and I'm in parallel working on just that.  The thought here was that
if some other process had started to manipulate the disk; then our read of the
size wouldn't be accurate.  However, thinking more on this; there's not much
else cloud-init can do if something else makes changes outside of it's control.
As such, I've dropped the locking added in get_size().

> 
> Secondly, I may be reading this wrong, but the added complexity of flock
> is just to improve the log message. I agree that this is useful
> information. Growpart writes (mostly) that information to stdout.
> Instead of reading it again, lets parse the output of the program
> that did the change:
> 
> CHANGED: partition=1 start=2048 old: size=1021952 end=1024000 new: size=2045919,end=2047967
> 
> It sucks that this output could be better formed. And that it is missing
> sector size, but it is definitely parseable with a fairly simple regex.
> I'd suggest we add sector size to growpart output, fix the lack of a ','
> between old size/new, and accept either one here.

I didn't quite parse what you meant here; but I eventually got there.
I've now implemented a get_size() for the ResizeGrowpart class which
uses regex to parse out the before/after values.  I've used your example
outputs; I'll check in on Xenial/Centos versions here to see if there's yet
another variant that needs handling.

Please take another look when you get a chance.

smoser commented:
> > This seems mostly "knee jerk" response, adding some code and extensive logging which has never been needed in 6 years.
> 
> It seems that things have changed, whether it be systemd, udev, locking, sfdisk, etc. If we had encountered a bug in the path, we would have wished we had better logging in cc_growpart to understand what code actually ran.
> 

This is always a trade off. It is worsened because cloud-init by default logs at DEBUG, and doesn't have anything louder than that.  But the result of "oh it would have been helpful if we had more logging" being applied without restraint is:

    2020-02-11 19:00:05,977 - util.py[DEBUG]: Reading from /var/lib/cloud/seed/nocloud/user-data (quiet=False)
    ...
    2020-02-11 19:00:05,977 - util.py[DEBUG]: Read 18 bytes from /var/lib/cloud/seed/nocloud-net/vendor-data

Now, forever more, we'll have logs on every system with verbose output of 'growpart' in order to facilitate the once-in-multiple-years occurance that someone wants to look at it.

Thats is what I mean by 'knee-jerk'. Anytime you're debugging it is useful to have more info on what you wanted to know.  But it is noise to ignore in all other cases.

smoser commented on cloudinit/config/cc_growpart.py:
```diff
@@ -139,25 +142,46 @@ def available(self):
-                util.logexc(LOG, "Failed growpart --dry-run for (%s, %s)",
-                            diskdev, partnum)
+                util.logexc(LOG, "Failed: growpart %s %s: %s",
+                            diskdev, partnum, err)
```
If you get here, then you probably did *not* log the stdout above, so you will have not logged that anywhere.

raharper replied:
Better to log the exception since ProcessExecutionError includes everything (cmd, stdout, stderr).

smoser commented on cloudinit/config/cc_growpart.py:
```diff
@@ -139,25 +142,46 @@ def available(self):
-            util.subp(["growpart", '--dry-run', diskdev, partnum])
+            out, err = util.subp(["growpart", '-v', diskdev, partnum])
+            LOG.debug('growpart: %s', out)
+            return self.get_size(out)
```
These two lines will *not* raise a ProcessExecutionError, so you should remove them from this `try`.
You should just do two try/excepts

    try:
        util.subp(...)
    except util.ProcessExecutionError as e:
        if e.exit_code != 1:
            util.logexc(LOG, "failed: growpart %s %s", diskdev, partnum)

    LOG.debug("growpart output: %s", out)
    try:
        return self.get_size(out)
    except ValueError as e:
        ...
You should verify that above... but the logexc should actually log the string representation of the ProcessExecutionError, which is *better* than your message here.  As that message would have stdout included, and error case, your you would not have ever logged stdout.  The ProcessExecutionError would happen before `LOG.debug('growpart: %s')` would have ran.

smoser commented on cloudinit/config/cc_growpart.py:
```diff
@@ -139,25 +142,46 @@ def available(self):
+        CHANGED: resizer=sfdisk partition=1 start=2048 old: size=1021952
+                 end=1024000 new: size=2045919 end=2047967
+        CHANGED: partition=1 start=2048 old: size=1021952 end=1024000
+                 new: size=2045919,end=2047967
```
This 'size' is in sectors. (I'm 95% sure of that)
Previously get_size() returned size in bytes.

So you have to figure out what the sector size is and multiply that.

Thats what I suggested adding to growpart.

raharper replied:
Good catch, I'll confirm and convert to bytes

smoser commented:
Oh.  I just realized/remembered.
Previously, we did not capture growpart output, and that was somewhat by design.
console output would include the CHANGED message.

Now you've changed that, so that it will not appear on the console.
So if something goes wrong, instead of being on console, this will only be seen in /var/log/cloud-init.log.

I do actually think single line CHANGED message on first boot to the console is useful.  (well more useful than the average of the 1000s of lines that are put there).  But feel free to tell me thats just my  personal noise.

raharper commented on cloudinit/config/cc_growpart.py:
```diff
@@ -139,25 +142,46 @@ def available(self):
+            util.logexc(LOG, "get_size error: %s", e)
             raise ResizeFailedException(e)
 
-        return (before, get_size(partdev))
```
In the NOCHANGE scenario; I'm seeing a stack trace; so I need to fix this before this can land.

raharper commented:
> Oh. I just realized/remembered.
> Previously, we did not capture growpart output, and that was somewhat by design.
> console output would include the CHANGED message.
> 
> Now you've changed that, so that it will not appear on the console.
> So if something goes wrong, instead of being on console, this will only be seen in /var/log/cloud-init.log.
> 
> I do actually think single line CHANGED message on first boot to the console is useful. (well more useful than the average of the 1000s of lines that are put there). But feel free to tell me thats just my personal noise.

I agree it's useful to see on console; and I think it should also be in the log.  I can use util.multi_log() to get them both places.

raharper commented:
> Oh. I just realized/remembered.
> Previously, we did not capture growpart output, and that was somewhat by design.
> console output would include the CHANGED message.

util.subp() captures stdout/stderr by default, so I don't think they made it to the console..

smoser commented on cloudinit/config/cc_growpart.py:
```diff
@@ -93,6 +93,10 @@ class RESIZE(object):
 
+# Linux hard-codes sector size to 512 for user-space
+# https://lkml.org/lkml/2015/8/17/269
+SECTOR_SIZE_SHIFT = 9  # 512
```
For reference, this is a 4K disk.

    # sfdisk -l /dev/sdb 
    Disk /dev/sdb: 2.2 TiB, 2398992990208 bytes, 585691648 sectors
    Units: sectors of 1 * 4096 = 4096 bytes
    Sector size (logical/physical): 4096 bytes / 4096 bytes
    I/O size (minimum/optimal): 65536 bytes / 65536 bytes
    Disklabel type: gpt
    Disk identifier: 9CD0A3AF-F802-40C8-A6E4-23E3355D72AD

    Device     Start       End   Sectors  Size Type
    /dev/sdb1    256 585691391 585691136  2.2T Linux LVM

raharper replied:
> No it doesn't.
> Growpart is reading sfdisk output. And sfdisk is outputting sectors. Those are 512 or 4096.
> 
> I can try tomorrow, I have some 4k disks, but you can also text in curtin-like vm.

It actually *does* but for it's sysfs 'size' file;

At this point, if I'm opening a file, we can just keep the original get_size();  While I can read sysfs here and I'm reasonable sure that growpart is only used on Linux path (where gpart is the bsd path);  the open + seek is pretty straight forward; more so than the parsing of the growpart output which isn't always the same between versions.

Given changes to growpart to emit the size in bytes (or sector_size );  I'd be OK with parsing this;  without a new growpart, I do think the original get_size() is more robust than the parsing of older growpart releases.   

If the growpart change lands, we can submit a new PR to detect the --size-in-bytes flag and then enable parsing of growpart output in lieu of using the get_size() open + seek.

smoser replied:
> If the growpart change lands, we can submit a new PR to detect the --size-in-bytes flag and then enable parsing of growpart output in lieu of using the get_size() open + seek.

I'm fine with keeping the old mechanism in place.

raharper commented on cloudinit/config/cc_growpart.py:
```diff
@@ -139,24 +146,53 @@ def available(self):
 
+        # Convert sectors to bytes.  growpart reports size in sectors
+        if before:
+            before = str(int(before) << SECTOR_SIZE_SHIFT)
```
I don't find it odd to expect shifting to be faster.  However, it's not in python.
https://stackoverflow.com/questions/37053379/times-two-faster-than-bit-shift-for-python-3-x-integers

raharper commented on cloudinit/config/cc_growpart.py:
```diff
@@ -139,24 +146,53 @@ def available(self):
+                                  rcs=[0, 1])
         except util.ProcessExecutionError as e:
-            util.logexc(LOG, "Failed: growpart %s %s", diskdev, partnum)
+            util.logexc(LOG, "Failed: growpart error: %s", e)
```
Right
